1-Draft/RFCNNNN-Improve-Background-Operator.md

```diff
@@ -0,0 +1,216 @@
+checklist item that is partially checked off than a feature that adds
+significant, tangible value to the PowerShell community. In a nutshell, we can
+and should do better.
+
+Note that there is a [related RFC about adding a `~&` ThreadJob operator](https://github.com/PowerShell/PowerShell-RFC/pull/205).
+
+## Motivation
+
+As a user,<br/>
+I can reference the most recent job(s) launched with `Start-Job` or the `&` background operator in a `$!` variable<br/>
+So that I can write my handling of those jobs in a consistent way without the job object or id, using the same variable that is used in bash for the same purpose.
+
+As a user,<br/>
+I can launch jobs with a new `&!` background operator that will show job progress in an overlay<br/>
+So that watch the concurrent progress of multiple background jobs in my host without blocking my console.
+
+As a user,<br/>
+I can use the `&` operator (and the proposed `&!` operator) after the stop parsing sigil (`--%`)<br/>
+So that I can launch legacy commands in the background more easily, without having to use `Start-Job`.
+
+Also, for completeness/functional parity with cmdlets:
+
+As a user,<br/>
+I can see the output of jobs launched with `Start-Job` or `Start-ThreadJob` in my current host by using a new `-ShowOutput` switch<br/>
+So that I can launch multiple background jobs and watch their concurrent progress in my host without blocking my console.
+
+As a user,<br/>
+I can show or hide the output of running jobs by using new `Show-JobOutput` and `Hide-JobOutput` cmdlets<br/>
+So that I can control the display and monitoring of running jobs without having to muck around with `Receive-Job -Keep`.
+
+As a PowerShell contributor,<br/>
+I can see the process ID associated with any background job by looking at the job members<br/>
+So that I can more easily attach the Visual Studio debugger to that job when I need to.
+
+Last, one describing the existing motivation for the `&` background operator
+(this is already in place today):
+
+As a user,<br/>
+I can launch jobs silently with the `&` background operator<br/>
+So that I can easily run pipelines as jobs in my automation solutions without showing output.
+
+## User Experience
+
+### Launching a background job with the output visible in the host
+
+```powershell
+# This command:
+Get-Process &!
+# Would return the same output as this command:
+Start-Job {Get-Process} -ShowOutput
+```
+
+```output
+Id     Name            PSJobTypeName   State         HasMoreData     Location             Command
+--     ----            -------------   -----         -----------     --------             -------
+1      Job1            BackgroundJob   Running       True            localhost            Microsoft.PowerShell.Man…
+```
+
+In addition to the output above, which is the job object created by the command
+that was invoked, PowerShell would show the output from the job (and any other
+job) launched this way in an overlay much like the output from Write-Progress,
```

Comment:
Oh, and I forgot we don't even support the overlay on non-Windows, so it's kind of a non-starter there too...

Reply:
Maybe. That's what bash does (just sends output to the terminal, when a command is invoked with `&` at the end). I was looking at `Write-Progress` and considering an overlay so that you wouldn't have the output messing up your prompt, but I didn't know the overlay doesn't work on non-Windows (shows how much I use `Write-Progress`). I wonder if it could be targeted, and simply sent to the host so that the default host could just output it in the terminal, or another host could send it into a separate window, or a different pane if the host supports multiple windows, etc.

Reply:
Have you thought about something like a `Receive-Job -Tail` or something like that that would "enter" a job and attach it's output stream to the current runspace? Or an `Enter-Job` so you could pop the session stack back out to your previous interactive session.

Reply:
These are a really interesting ideas. They immediately me wonder if `Watch` is a verb, and it is, so we could have `Watch-Job` for anyone who is wondering what is going on with a job. Seems appropriate since it would be in your current console, which is in contrast to `Show` would would imply opening a new window to see the job in progress.

I prefer `Watch` over `Enter` because I would like to be able to monitor multiple jobs at once.

With this we could have `Watch-Job` to start watching one or more jobs, and `Watch-Job -Off` to stop watching one or more jobs.

Reply:
Feedback incorporated in an update that removes the overlay idea in favor of integrating monitoring output in the console. This update also replaces the `Show/Hide-JobOutput` cmdlets with `Watch-Job`/`Watch-Job -Off` (more discoverable, and a better verb for inline monitoring). If you feel I missed any point here, let me know.

Reply:
@KirkMunro don't think Watch is, but Show-Job might be an OK one to start with? 🙂

Reply:
@vexx32 [`Watch` is an approved verb](https://docs.microsoft.com/en-us/dotnet/api/System.Management.Automation.VerbsCommon.Watch?view=pscore-6.2.0), specifically created for monitoring needs like this. From the verb guidelines:

|Verb (alias)|Action|
|--|--|
|Watch (wc)|Continually inspects or monitors a resource for changes.|